tycho-client/src/feed/synchronizer.rs

```diff
@@ -571,10 +468,92 @@ where
+                                                        entrypoints: snapshot_response.traced_entry_points.get(&component.id).cloned().unwrap_or_default(),
+                                                    },
+                                                ))
+                                            } else if component_ids.contains(&component.id) {
+                                                error!(component_id = ?component.id, "Missing state for native component!");
+                                                None
+                                            } else {
+                                                None
+                                            }
+                                        }).collect();
+
+                                        let contract_states = snapshot_response.vm_storage;
+                                        trace!(states=?&contract_states, "Retrieved ContractState");
+
+                                        let contract_address_to_components = self.component_tracker.components.iter()
+                                            .filter_map(|(id, comp)| {
+                                                if component_ids.contains(id) {
+                                                    Some(comp.contract_ids.iter().map(|address| (address.clone(), comp.id.clone())))
+                                                } else {
+                                                    None
+                                                }
+                                            })
+                                            .flatten()
+                                            .fold(HashMap::<Bytes, Vec<String>>::new(), |mut acc, (addr, c_id)| {
+                                                acc.entry(addr).or_default().push(c_id);
+                                                acc
+                                            });
+
+                                        let vm_storage = contract_ids.iter().filter_map(|address| {
+                                            if let Some(state) = contract_states.get(address) {
+                                                Some((address.clone(), state.clone()))
+                                            } else if let Some(ids) = contract_address_to_components.get(address) {
+                                                error!(?address, ?ids, "Component with lacking contract storage encountered!");
+                                                None
+                                            } else {
+                                                None
+                                            }
+                                        }).collect();
+
+                                        Snapshot { states, vm_storage }
+                                    }
+                                };
```

Comment:
This seems very similar to the code above. I really dislike having such complex logic duplicated.